src/libutil/filesystem.cpp

```diff
@@ -514,6 +523,182 @@ Filesystem::open (std::ofstream &stream, string_view path,
+        f |= _O_BINARY;
+    } else {
+        f |= _O_TEXT;
+    }
+    return f;
+}
+
+static std::istream*
+open_ifstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+    int fd;
+    int oflag = ios_open_mode_to_oflag(mode);
+    errno_t errcode = _wsopen_s(&fd, wpath.c_str(), oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE);
+    if (errcode != 0) {
+        return 0;
+    }
+    __gnu_cxx::stdio_filebuf<char>* buffer = new __gnu_cxx::stdio_filebuf<char>(fd, mode, 1);
+    if (!buffer) {
+        return 0;
+    }
+    return new std::istream(buffer);
+}
+
+static std::ostream*
+open_ofstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+    int fd;
+    int oflag = ios_open_mode_to_oflag(mode);
+    errno_t errcode = _wsopen_s(&fd, wpath.c_str(), oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE);
+    if (errcode != 0) {
+        return 0;
+    }
+    __gnu_cxx::stdio_filebuf<char>* buffer = new __gnu_cxx::stdio_filebuf<char>(fd, mode, 1);
+    if (!buffer) {
+        return 0;
+    }
+    return new std::ostream(buffer);
+}
+#else // MSVC or Unix
+
+#ifdef _WIN32
+#ifndef _MSC_VER_
+#error "open_ifstream_impl only supports GCC or MSVC"
+#endif
+#endif
+
+static std::ifstream*
+open_ifstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    
+#ifdef _WIN32
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+#endif
+    std::ifstream *ret = new std::ifstream();
+    if (!ret) {
+        return 0;
+    }
+    try {
+        ret->open(
+#ifdef _WIN32
+                  wpath.c_str(),
+#else
+                  path.c_str(),
+#endif
+                  mode);
+    } catch (const std::exception & e) {
+        delete ret;
+        return 0;
+    }
+    
+    if (!*ret) {
+        delete ret;
+        return 0;
+    }
+    
+    return ret;
+} // open_ifstream_impl
+
+static std::ofstream*
+open_ofstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+#ifdef _WIN32
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+#endif
+    std::ofstream *ret = new std::ofstream();
+    if (!ret) {
+        return 0;
+    }
+    try {
+        ret->open(
+#ifdef _WIN32
+                  wpath.c_str(),
+#else
+                  path.c_str(),
+#endif
+                  mode);
+    } catch (const std::exception & e) {
+        delete ret;
+        return 0;
+    }
+    
+    if (!*ret) {
+        delete ret;
+        return 0;
+    }
+    
+    return ret;
+}
+#endif //#if defined(_WIN32) &&  defined(__GLIBCXX__)
+
+
+void
+Filesystem::open (std::istream** stream,
+                  string_view path,
+                  std::ios_base::openmode mode)
+{
+    if (!stream) {
+        return;
+    }
+    *stream = open_ifstream_impl(path, mode | std::ios_base::in);
+    if (mode & std::ios_base::ate) {
```

Comment:
I think you need to handle open_ifstream_impl failing to open the file and setting *stream to NULL.